services/typeddata/sign.go

```diff
@@ -0,0 +1,51 @@
+package typeddata
+
+import (
+	"crypto/ecdsa"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+const (
+	chainIDKey = "chainId"
+)
+
+var (
+	// x19 to avoid collision with rlp encode. x01 version byte defined in EIP-191
+	messagePadding = []byte{0x19, 0x01}
+)
+
+func hashToSign(typed TypedData) (rst common.Hash, err error) {
```

Comment:
I found it a bit hard to map this function to the EIP doc, and confused at the beginning with `hashStruct`. Seems like `hashToSign` is what described in spec as

```
encode(domainSeparator : 𝔹²⁵⁶, message : 𝕊) = "\x19\x01" ‖ domainSeparator ‖ hashStruct(message)
```

So maybe it can be renamed to `encode`, and, even better, be a method of `TypedData`? So the usage is:

```go
hash, err := typed.encode()
```

Reply:
i dont like wrapping as method because this code depends on several constants. message padding,  key for domain separator. they can change. and same struct will have to support multiple variants just because message padding changed